### PR TITLE
Reject SmartPointer constructions not serving the purpose

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
Tracking issue: #123430

With this PR we will reject a row of malformed `SmartPointer` implementor candidates.

cc @Darksonn @davidtwco for context.

### CODE REVIEW

File: tests/ui/deriving/deriving-smart-pointer.rs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 use std::marker::SmartPointer;
 
 #[derive(SmartPointer)]
+#[repr(transparent)]
 struct MyPointer<'a, #[pointee] T: ?Sized> {
     ptr: &'a T,
 }
```
Comment (Darksonn) [inline tests/ui/deriving/deriving-smart-pointer.rs:10]: Does it still work if you swap the order?
```rs
#[repr(transparent)]
#[derive(SmartPointer)]
struct MyPointer<'a, #[pointee] T: ?Sized> {
    ptr: &'a T,
}
```

Comment (dingxiangfei2009) [reply tests/ui/deriving/deriving-smart-pointer.rs:10]: Yes, and a test is just now added to certify it.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**